ledger/eval.go

```diff
@@ -1218,39 +1221,188 @@ func eval(ctx context.Context, l ledgerForEvaluator, blk bookkeeping.Block, vali
-	defer wg.Done()
-	maybelookup := func(addr basics.Address) {
-		if addr.IsZero() {
-			return
+// loadedTransactionGroup is a helper struct to allow asyncronious loading of the account data needed by the transaction groups
+type loadedTransactionGroup struct {
+	// group is the transaction group
+	group []transactions.SignedTxnWithAD
+	// balances is a list of all the balances that the transaction group refer to and are needed.
+	balances []basics.BalanceRecord
+	// err indicates whether any of the balances in this structure have failed to load. In case of an error, at least
+	// one of the entries in the balances would be uninitialized.
+	err error
+}
+
+// loadAccounts loads the account data for the provided transaction group list. It also loads the feeSink account and add it to the first returned transaction group.
+// The order of the transaction groups returned by the channel is identical to the one in the input array.
+func loadAccounts(ctx context.Context, l ledgerForEvaluator, rnd basics.Round, groups [][]transactions.SignedTxnWithAD, feeSinkAddr basics.Address) chan loadedTransactionGroup {
+	outChan := make(chan loadedTransactionGroup, len(groups))
+	go func() {
+		// groupTask helps to organize the account loading for each transaction group.
+		type groupTask struct {
+			// balances contains the loaded balances each transaction group have
+			balances []basics.BalanceRecord
+			// balancesCount is the number of balances that nees to be loaded per transaction group
+			balancesCount int
+			// done is a waiting channel for all the account data for the transaction group to be loaded
+			done chan error
 		}
-		state.lookup(addr)
-	}
-	for _, stxn := range payset {
-		select {
-		case <-ctx.Done():
-			return
-		default:
+		// addrTask manage the loading of a single account address.
+		type addrTask struct {
+			// account address to fetch
+			address basics.Address
+			// a list of transaction group tasks that depends on this address
+			groups []*groupTask
+			// a list of indices into the groupTask.balances where the address would be stored
+			groupIndices []int
 		}
-		state.lookup(stxn.Txn.Sender)
-		maybelookup(stxn.Txn.Receiver)
-		maybelookup(stxn.Txn.CloseRemainderTo)
-		maybelookup(stxn.Txn.AssetSender)
-		maybelookup(stxn.Txn.AssetReceiver)
-		maybelookup(stxn.Txn.AssetCloseTo)
-		maybelookup(stxn.Txn.FreezeAccount)
-		for _, xa := range stxn.Txn.Accounts {
-			maybelookup(xa)
+		defer close(outChan)
+
+		accountsChannels := make(map[basics.Address]*addrTask)
+		addressesCh := make(chan *addrTask, len(groups)*16*10)
+		// totalBalances counts the total number of balances over all the transaction groups
+		totalBalances := 0
+		initAccount := func(addr basics.Address, wg *groupTask) {
+			if addr.IsZero() {
+				return
+			}
+			if task, have := accountsChannels[addr]; !have {
+				task := &addrTask{
+					address:      addr,
+					groups:       make([]*groupTask, 1, 4),
+					groupIndices: make([]int, 1, 4),
```

Comment:
why do you set `cap=4`?

Reply:
I'm not a big fan of this part either. Let me see if I preallocate and sub-allocate that.

Reply:
Preallocating wasn't as trivial as I initially thought; I'll defer that to a separate task.
The capacity = 4 is a good fit for most of the trivial use cases. using just cap=1 is usually not a smart move, since there is almost always more than a single account involved.

Reply:
ok, this is just a guess, not exact number, good enough